manifests/server.pp

```diff
@@ -0,0 +1,119 @@
+# @summary Configure backup server
+#
+# @param additional_sshd_match_params
+#   Hash of additional sshd match parameters.
+#   Passed to sshd::allow_from defined type from profile_backup::allow_client_on_server.
+#
+# @param backup_directory
+#   Directory path where backups are stored for each client.
+#
+# @param clients
+#   Clients that need to be manually configured.
+#   Backup clients that are using the same Puppet server should not need to
+#   be added here as those are added via 'exported resources' via PuppetDB.
+#   This is a hash that contains all the parameters for `allow_client_on_server`:
+#   ```yaml
+#   profile_backup::server::clients:
+#     client1:
+#       hostname: "client1.local"
+#       ip: "172.1.2.3"
+#       key: "AAAAB..."  # PUBLIC KEY
+#       key_type: "ssh-rsa"  # ENCRYPTION TYPE
+#   ```
+#
+# @param gid
+#   Group id of user that owns backup files.
+#
+# @param groupname    
+#   Groupname that owns backup files.
+#
+# @param require_storage
+#   Optionally list resources (e.g., services) that require storage
+#   in order to function.
+#
+# @param storage_dependencies
+#   Optionally list resources (e.g. mounts) that should be present before
+#   setting up backup resources on server. Should be in the form that would
+#   be specified as a requirement ("before") various backup server
+#   resources, e.g.:
+#     Lvm::Logical_volume::backups
+#
+# @param uid
+#   User id of user that owns backup files.
+#
+# @param username
+#   Username that owns backup files and allowed access.
+#
+# @example
+#   include profile_backup::server
+class profile_backup::server (
+  Hash   $additional_sshd_match_params,
+  String $backup_directory,
+  Hash   $clients,
+  String $gid,
+  String $groupname,
+  Array[String] $require_storage,
+  Array[String] $storage_dependencies,
+  String $uid,
+  String $username,
+) {
+  include profile_backup::common
+
+  # FLEXIBLE REQUIRE LOGIC TO ENSURE $backup_directory IS MOUNTED
+  #   FLEXIBLE TO SUPPORT LUSTRE, NFS, BINDMOUNT, ETC.
+  #   FOLLOWING IS COPIED FOR profile_slurm
+  #   BUT NOT QUITE WHAT WE NEED HERE YET
+  #   NEED THE STORAGE REQUIREMENTS TIED TO allow_client_on_server DEFINED TYPE
+  $storage_dependencies.each | $dependency | {
+    $require_storage.each | $dependent | {
+      $dependency -> $dependent
+    }
+  }
```

Comment:
what is this logic attempting?

Reply:
This is the storage dependency logic copied from `profile_slurm`.

As noted in the comments it does not work for what we need. There is no 'service' or simple resource that we need to setup dependencies for. But rather we need a generalized way to make sure that the `profile_backup::server::backup_directory` is mounted before the `allow_client_on_server` defined type commands are called later in `profile_backup::server`.

I just have these here as reminders that we need to address this somehow eventually.

Reply:
https://github.com/ncsa/puppet-profile_backup/blob/1164d9b296e086290fb51cce784ebeee8fcc9db1/manifests/server.pp#L88-L101 on the latest commit hopefully illustrates and documents better what we're looking for here.

Reply:
I have replaced this with with `$profile_backup::server::allow_client_requires` which is then referenced in the `profile_backup::server::allow_client` defined type for the 2 specific resources there that optionally be needing these requirements to make sure they are writing to the correct disk, etc.

This is much simpler and hopefully more obvious with what is going on.